domain/application/state/unitstate.go

```diff
@@ -0,0 +1,1413 @@
+	// First see if there's an existing address recorded.
+	err = tx.Query(ctx, selectAddressUUIDStmt, ipAddr).Get(&ipAddr)
+	if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+		return errors.Errorf("querying existing cloud container address for device %q: %w", deviceUUID, err)
+	}
+
+	// Create a UUID for new addresses.
+	if errors.Is(err, sqlair.ErrNoRows) {
+		addrUUID, err := uuid.NewUUID()
+		if err != nil {
+			return errors.Capture(err)
+		}
+		ipAddr.AddressUUID = addrUUID.String()
+	}
+
+	// Update the address values.
+	if err = tx.Query(ctx, upsertAddressStmt, ipAddr).Run(); err != nil {
+		return errors.Errorf("updating cloud container address attributes for device %q: %w", deviceUUID, err)
+	}
+	return nil
+}
+
+func (st *InsertIAASUnitState) upsertCloudContainerPorts(ctx context.Context, tx *sqlair.TX, unitUUID string, portValues []string) error {
+	type ports []string
+
+	ccPort := unitK8sPodPort{
+		UnitUUID: unitUUID,
+	}
+	deleteStmt, err := st.Prepare(`
+DELETE FROM k8s_pod_port
+WHERE port NOT IN ($ports[:])
+AND unit_uuid = $unitK8sPodPort.unit_uuid;
+`, ports{}, ccPort)
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	upsertStmt, err := st.Prepare(`
+INSERT INTO k8s_pod_port (*)
+VALUES ($unitK8sPodPort.*)
+ON CONFLICT(unit_uuid, port)
+DO NOTHING
+`, ccPort)
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	if err := tx.Query(ctx, deleteStmt, ports(portValues), ccPort).Run(); err != nil {
+		return errors.Errorf("removing cloud container ports for %q: %w", unitUUID, err)
+	}
+
+	for _, port := range portValues {
+		ccPort.Port = port
+		if err := tx.Query(ctx, upsertStmt, ccPort).Run(); err != nil {
+			return errors.Errorf("updating cloud container ports for %q: %w", unitUUID, err)
+		}
+	}
+
+	return nil
+}
+
+func (st *InsertIAASUnitState) k8sSubnetUUIDsByAddressType(ctx context.Context, tx *sqlair.TX) (map[network.AddressType]string, error) {
+	result := make(map[network.AddressType]string)
+	subnetStmt, err := st.Prepare(`
+ SELECT &subnet.*
+ FROM subnet
+ `, subnet{})
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	var subnets []subnet
+	if err = tx.Query(ctx, subnetStmt).GetAll(&subnets); err != nil {
+		return nil, errors.Errorf("getting subnet uuid: %w", err)
+	}
+	// Note: Today there are only two k8s subnets, which are a placeholders.
+	// Finding the subnet for the ip address will be more complex
+	// in the future.
+	if len(subnets) != 2 {
+		return nil, errors.Errorf("expected 2 subnet uuid, got %d", len(subnets))
+	}
+
+	for _, subnet := range subnets {
+		addrType := addressTypeForUnspecifiedCIDR(subnet.CIDR)
+		result[addrType] = subnet.UUID
+	}
+	return result, nil
+}
+
+// ensureFutureUnitNetNode exists to ensure that a netnode uuid that is about to
+// be used for a machine exists. We do this because the business logic around if
+// netnode uuid for a unit is shared or already exists is outside the scope of
+// state.
+func (st *InsertIAASUnitState) ensureFutureUnitNetNode(
```

Comment:
Not sure about the naming of the this method, but we can come back around to this.